styx-scheduler-service/src/main/java/com/spotify/styx/Scheduler.java

```diff
@@ -159,17 +173,44 @@ void tick() {
 
     timedOutInstances.forEach(this::sendTimeout);
 
-    for (InstanceState eligibleInstance : eligibleInstances) {
-      final boolean proceed = limitAndDequeue(
-          resources, workflowResourceReferences, currentResourceUsage, eligibleInstance);
-      if (!proceed) {
-        break;
-      }
-    }
+    limitAndDequeue(config, resources, workflowResourceReferences, currentResourceUsage,
+        eligibleInstances);
 
     updateStats(resources, currentResourceUsage);
   }
 
+  private void limitAndDequeue(
+      StyxConfig config,
+      Map<String, Resource> resources,
+      Map<WorkflowId, Set<String>> workflowResourceReferences,
+      Map<String, Long> currentResourceUsage, List<InstanceState> eligibleInstances) {
+
+    // Enable gating unless disabled in runtime config
+    final WorkflowExecutionGate gate = config.executionGatingEnabled() ? this.gate : NOOP;
+
+    // Process the eligible instances in batches in order to parallelize execution blocker lookup
+    for (List<InstanceState> batch : Lists.partition(eligibleInstances, SCHEDULING_BATCH_SIZE)) {
+
+      // Asynchronously look up execution blockers for a batch of instances
+      final Map<WorkflowInstance, CompletionStage<Optional<ExecutionBlocker>>> executionBlockers =
+          batch.stream()
+              .map(InstanceState::workflowInstance)
+              .collect(toMap(identity(), gate::executionBlocker));
+
+      // Evaluate each instance in the batch for dequeuing
+      for (InstanceState instance : batch) {
+        final boolean proceed = limitAndDequeue(
+            resources, workflowResourceReferences, currentResourceUsage, instance,
+            executionBlockers.getOrDefault(instance.workflowInstance(), NO_BLOCKER));
```

Comment:
This map is built by iterating through the batch, in what case we might get the default? It doesn't seem necessary to use a map. Do I misunderstand something?

Reply:
The reason for storing the blockers in a separate collection is to look them up in parallel for the entire batch.

Perhaps I would have used a list instead of a map if java offered a nice way of iterating over two lists. E.g. zipping. But alas.

An alternative here would be to change this for loop into an old-school for loop with an integer counter.

Reply:
Ok old school was the also the alternative coming to my mind. ;)

Anyway getOrDefault doesn’t seem to be necessary and can be confusing when reading the code after a while.

Reply:
Went for old-school for loop, feels cleaner 👍